File: Population.java


public class Population {

    Chromosome[] chro;
    Chromosome[] fitchro;
    private static final double crossRate = 0.5;
    private static final double mutationRate = 0.02;
    private static final int selectRate = 5;
    private static final int mutationFactor = 10;

    public Population(int pop) {
        chro = new Chromosome[pop];
        fitchro = new Chromosome[selectRate];
    }

    public Chromosome get(int num) {return chro[num];}
    public void set(int num, Chromosome c){chro[num]=c;}
    public int size() {return chro.length;}

    public void initialize(){
    	for (int i = 0; i < size(); i++) {
        	Chromosome c = new Chromosome();
                c.generate(c);
                set(i, c);
        }
    }
    public void populate(){
    	for (int i = 0; i < size(); i++) {
    		Chromosome c= new Chromosome();
    		c.generate(fitchro[(int) Math.round(Math.random())*(selectRate-1)]);
            set(i, c);
        }
    }
    public void selection(Population pop){
       
        for(int u=0;u<selectRate;u++){
        	fitchro[u]=getFittest();
        	getFittest().kill();
        }
    }
    public void mutation() {
    	for(Chromosome c: chro){
	    	if (Math.random() <= mutationRate) {
	    		c.setx(c.getx() + (int) Math.round(Math.random()*mutationFactor));
	        }
	    	if (Math.random() <= mutationRate) {
	    		c.sety(c.gety() + (int) Math.round(Math.random()*mutationFactor));
	        }
	    	if (Math.random() <= mutationRate) {
	    		c.setx(c.getx() - (int) Math.round(Math.random()*mutationFactor));
	        }
	    	if (Math.random() <= mutationRate) {
	    		c.sety(c.gety() - (int) Math.round(Math.random()*mutationFactor));
	        }
    	}
    }
    public void crossover() {
        for (int i = 0; i < selectRate; i++) {
        	for (int j = 0; j < selectRate; j++) {
		        if (Math.random() <= crossRate) {
		        	int u = fitchro[i].getx();
		        	fitchro[i].setx(fitchro[j].getx());
		        	fitchro[j].setx(u);	
		        }
		        if (Math.random() <= crossRate) {
		        	int u = fitchro[i].gety();
		        	fitchro[i].sety(fitchro[j].gety());
		        	fitchro[j].sety(u);	
		        }
        	}
        }
    }
    public int evaluation(){
    	return getFittest().getFitness();
    }
    public Chromosome getFittest() {
        Chromosome fittest = chro[0];
        for (int i = 0; i < size(); i++) {
            if (fittest.getFitness() <= get(i).getFitness()) {
                fittest = get(i);
            }
        }
        return fittest;
    }
}

File: Population.kt
import kotlin.math.roundToInt

class Population(pop: Int) {
    private var population = (0 until pop).map {
        Chromosome().apply { generateFrom(this) }
    }
    private var survivors = Array(selectRate) { Chromosome() }

    fun populate() {
        population.forEach { it.generateFrom(survivors.random()) }
    }

    fun selection() {
        (0 until selectRate).forEach { u ->
            survivors[u] = fittest
            fittest.kill()
        }
    }

    fun mutation() {
        population.forEach { c ->
            if (Math.random() <= mutationRate) {
                c.x += getMutationAmount()
            }
            if (Math.random() <= mutationRate) {
                c.y += getMutationAmount()
            }
            if (Math.random() <= mutationRate) {
                c.x -= getMutationAmount()
            }
            if (Math.random() <= mutationRate) {
                c.y -= getMutationAmount()
            }
        }
    }

    private fun getMutationAmount() = ((Math.random() * mutationFactor)).roundToInt()

    fun crossover() {
        for (i in 0 until selectRate) {
            for (j in 0 until selectRate) {
                if (Math.random() <= crossRate) {
                    exchangeX(survivors[i], survivors[j])
                }
                if (Math.random() <= crossRate) {
                    exchangeY(survivors[i], survivors[j])
                }
            }
        }
    }

    private fun exchangeY(first: Chromosome, second: Chromosome) {
        val u = first.y
        first.y = second.y
        second.y = u
    }

    private fun exchangeX(first: Chromosome, second: Chromosome) {
        val u = first.x
        first.x = second.x
        second.x = u
    }

    fun evaluation(): Int = fittest.fitness

    val fittest: Chromosome
        get() {
            return population.onEach {
                if (it.fitness == 0) {
                    val i0 = it.x - 3750
                    val i1 = it.y - 5000
                    val i2 = 570 * it.x
                    val i3 = 1500 * it.y + 25
                    it.fitness = -(i0 * i0) + i2 + -2 * (i1 * i1) + i3
                }
            }.maxByOrNull { it.fitness }!!
        }

    companion object {
        private const val crossRate = 0.5
        private const val mutationRate = 0.02
        private const val selectRate = 5
        private const val mutationFactor = 10
    }
}